### PR TITLE
Add a /config endpoint

### LINKED ISSUE #394: Add admin endpoints to introspect proxy config
It would be very useful for operations to be able to be able to view the current config that a proxy is running at a given point in time.

Let's add two admin endpoints /clusters and /filterchain that returns a simplified JSON of the current cluster and filterchain respectively back to the caller. e.g:

```
$ curl quilkin:8080/clusters
[{
  "name": "default-cluster",
  "endpoints": [{
    "address": {
      "socketAddress": {
          "address": "127.0.0.1",
          "portValue": 1234
      }
    }
  }]
]}
```

```
$ curl quilkin:8080/filterchain
{
  "filters": [{
    {
      "name": "quilkin.extensions.filters.debug.v1alpha1.Debug",
      "config": {
        id: "hello"
      }  
    }  
  }]
}
```

### PR DESCRIPTION
**Pull request by iffyio**
That returns a JSON of the current config the proxy is running with

This includes a breaking change, as the FilterFactory to now needs to return
the config in addition to the filter it created.

Resolves #394

### CODE REVIEW

**Comment by iffyio** [comment]
example output
```yaml
{
  "clusters": [
    {
      "name": "default-quilkin-cluster",
      "endpoints": [
        {
          "address": "127.0.0.1:4321",
          "metadata": {
            "quilkin.dev": {
              "tokens": [
                "AA=="
              ]
            }
          }
        }
      ]
    }
  ],
  "filterchain": {
    "filters": [
      {
        "name": "quilkin.extensions.filters.capture_bytes.v1alpha1.CaptureBytes",
        "config": {
          "metadataKey": "quilkin.dev/captured_bytes",
          "remove": true,
          "size": 1,
          "strategy": "SUFFIX"
        }
      },
      {
        "name": "quilkin.extensions.filters.token_router.v1alpha1.TokenRouter",
        "config": {
          "metadataKey": "quilkin.dev/captured_bytes"
        }
      }
    ]
  }
}
```

**Comment by iffyio** [comment]
Also this has a bunch of conflicts with #395 in the custom filter doc at least, I'll resolve them after that's merged

File: src/config/config_type.rs
Comment (XAMPPRocky) [inline src/config/config_type.rs:104]: If you take it by reference here, you don't need to pass it through the function and back to the caller, and we can also remove level of serialisation.

```suggestion
    fn get_json_config<T>(filter_name: &str, config: &T) -> Result<serde_json::Value, Error>
    where
        T: serde::Serialize + for<'de> serde::Deserialize<'de>,
    {
        serde_json::to_value(config.clone())
            .map_err(|err| {
                Error::DeserializeFailed(format!(
                    "filter `{}`: failed to serialize config to json: {}",
                    filter_name,
                    err.to_string()
                ))
            })
```

File: src/config/config_type.rs
Comment (XAMPPRocky) [inline src/config/config_type.rs:37]: Very minor nit: The "Marshall" terminology isn't what you'd use in Rust. You'd normally just say "serialisation/deserialisation", I guess it's because in Rust they're not really objects in the same way they are in Go or Python. I don't think it's too important though.

File: src/filters/factory.rs
Comment (XAMPPRocky) [inline src/filters/factory.rs:30]: Since most of the time we create a new a filter the configuration doesn't change, as a future improvement, we could use `Arc` here instead to share memory across filter instances.

File: src/filters/chain.rs
Comment (XAMPPRocky) [inline src/filters/chain.rs:131]: Along with changing the type to `FilterInstance` I would change these to call the top level variable `instance`. 
```suggestion
                    histogram.observe_closure_duration(|| instance.filter.read(ctx))?,
```

File: src/config/config_type.rs
Comment (XAMPPRocky) [inline src/config/config_type.rs:56]: This doesn't have to be in this PR, but I would like to combine these two type parameters into single trait, so that we can enforce it as part of the type system. Then you also only need to provide a single parameter.

```rust
trait FilterConfig {
   type Dynamic: prost::Message + Default;
   type Static: serde::Serialize
             + for<'de> serde::Deserialize<'de>
             + TryFrom<Self::Dynamic, Error = ConvertProtoConfigError>;
}

// src/filters/debug.rs
impl FilterConfig for Config {
    type Static = Self;
    type Dynamic = ProtoDebug;
}
```

File: src/filters/factory.rs
Comment (XAMPPRocky) [inline src/filters/factory.rs:35]: I'm not sure about this `From` implementation, if we add new things to `CreatedFilter` we'll have to change it, and it feels odd that `(Value, Box)` would work but `(Box, Value)` won't. What would you think about having a `new` method instead?

File: src/proxy/admin.rs
Comment (XAMPPRocky) [inline src/proxy/admin.rs:100]:
```suggestion
        (&Method::GET, "/config") => {
```

**Review by XAMPPRocky — COMMENTED** [review]
Looks good to me, mostly just minor naming concerns, some code changes, and some potential improvements that can come later.

**Comment by markmandel** [comment]
Since #395 is about ready to be merged (I need to make a small change, and it's good to go) -- since this is a particularly useful feature, should we wait until it's merged and bundle it into the 0.2.0 release?

**Comment by iffyio** [comment]
> Since #395 is about ready to be merged (I need to make a small change, and it's good to go) -- since this is a particularly useful feature, should we wait until it's merged and bundle it into the 0.2.0 release? (I'm leaning towards "yes", personally)

yes please :pray: also so we can avoid using a custom branch in our deployment internally until the next release

**Comment by markmandel** [comment]
Added to the milestone!

File: docs/src/filters/writing_custom_filters.md
Comment (markmandel) [inline docs/src/filters/writing_custom_filters.md:210]: I really like this being back in - since it slowly introduces people to the config option, and then we can move into `deserialize`, which is the more complex topic. 👍🏻

**Review by markmandel — APPROVED** [review]
Looks great to me!

If @XAMPPRocky is also happy, I'm very happy to have this merged 🥳

**Review by XAMPPRocky — APPROVED** [review]

**Pull request merged**
